### DISCUSSION TITLE
Must I use /auth/password/request?

### DISCUSSION THREAD

**Discussion by glowkeeper** [directus/directus#4549] [Archive]
Is pinging _/auth/password/request_ the *only* way a user can reset their password? 

The whole 'email a user' workflow is somewhat problematic (for me) because HIGH COMMAND has requested our app' eventually authenticates via mobile phone numbers. Eek!

Anyway, for now, I can put them off...

Meanwhile, if I'm tied to _/auth/password/request_, what form does PASSWORD_RESET_URL_ALLOW_LIST take? Is it some JSON object thingy, or a comma-separated string (i.e. PASSWORD_RESET_URL_ALLOW_LIST="http://my.fab.production.webserver/user/password/reset, http://my.local.devserver/user/password/reset")? Or some other form?  I've tried a few ways, but the API eventually responds to my `fetch` request with some JSON parse error that I haven't yet quite fathomed...

**Comment by rijkvanzanten** [comment]
> Is pinging /auth/password/request the only way a user can reset their password?

It's the only "natively" supported way as of right now, yes. 

> The whole 'email a user' workflow is somewhat problematic (for me) because HIGH COMMAND has requested our app' eventually authenticates via mobile phone numbers. Eek!

😱 On a serious note though: You can update users through the UsersService directly in a custom endpoint, so it should be possible to make a custom reset password flow that relies on sending text messages with a code through Twilio or the likes. You'd effectively make a custom `/custom/phone-auth/request` and reset.

> Meanwhile, if I'm tied to /auth/password/request, what form does PASSWORD_RESET_URL_ALLOW_LIST take? Is it some JSON object thingy, or a comma-separated string (i.e. PASSWORD_RESET_URL_ALLOW_LIST="http://my.fab.production.webserver/user/password/reset, http://my.local.devserver/user/password/reset")? Or some other form?

`PASSWORD_RESET_URL_ALLOW_LIST` is a CSV of URLs you accept, so `http://my.fab.production.webserver/user/password/reset, http://my.local.devserver/user/password/reset` would be correct.

> I've tried a few ways, but the API eventually responds to my fetch request with some JSON parse error that I haven't yet quite fathomed...

This to me sounds like a problem in what you're POSTing to the API in order to reset, not necessarily the configuration it uses 🤔 Could it be that you have a typo in the POST body to `/request` ?

**Reply by glowkeeper** [reply]
> You can update users through the UsersService directly in a custom endpoint, so it should be possible to make a custom reset password flow that relies on sending text messages with a code through Twilio or the likes. You'd effectively make a custom /custom/phone-auth/request and reset.

Brilliant - I have that endpoint working and I'll take a peek at Twilio - thank you so much for the suggestion!

> Could it be that you have a typo in the POST body to /request

I don't _think_ so - it looks exactly like all the other POSTs I'm making to directus:
```
    const userData = {
      email: directusUser@foor.bar,
      reset_url: http://my.local.devserver/user/password/reset
    }

    fetch(http://direcus.url/password/request, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(userData)
    })
```

**Reply by rijkvanzanten** [reply]
If this is copy-pasted, there's quotes missing around these values:

```js
const userData = {
      email: directusUser@foor.bar,
      reset_url: http://my.local.devserver/user/password/reset
    }
```

that should be `"directusUser@foor.bar"` and `"http://my.local.devserver/user/password/reset"` respectively 

Same goes for the fetch:

`fetch(http://direcus.url/password/request, ...` should be `fetch("http://direcus.url/password/request", ...`

**Reply by glowkeeper** [reply]
Yeah - me bad :( That's not the code, that's me editing the comment - it is all quoted correctly, so indeed, it looks like this:
```
const userData = {
      email: "directusUser@foor.bar",
      reset_url: "http://my.local.devserver/user/password/reset"
    }

    fetch("http://direcus.url/password/request", {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(userData)
    })
    
```

**Reply by rijkvanzanten** [reply]
This snippet looks good to me 🤔 What is it sending up in the network tab?

**Reply by glowkeeper** [reply]
Here's a screengrab:

<img width="1440" alt="Screenshot 2021-03-16 at 18 08 17" src="https://user-images.githubusercontent.com/9264929/111358797-a9f1d200-8682-11eb-9d33-6bc274e1df61.png">

And here's the _actual_ code:

```
export const reset = (user: Reset) => {
  return async (dispatch: AppDispatch) => {

    let d = new Date(Date.now())
    let dateText = d.toString()

    const userData = {
      email: `${user.email}`,
      reset_url: `${web}${Local.resetPassword}`
    }

    const url = `${dbase}${Remote.passwordResetRequest}`
    //console.log("User: ", userData, url)

    fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(userData)
    })
    .then(response => {

      if (!response.ok) {
          const status = response.status
          const statusText = response.statusText
          return response.json()
          .then(data => {
              const txData = {
                code: status.toString(),
                summary: `${UserConfig.resetFailure}: ${statusText}`,
                time: `${dateText}`
              }
              throw new Error(statusText)
          })
      } else {

        console.log("did I get here?", response)
        return response.json()
      }
    })
    .then(data => {

      const txData = {
          code: "200",
          summary: `${UserConfig.resetSuccess}`,
          time: `${dateText}`
      }
      dispatch(write({data: txData})(TxActionTypes.TX_SUCCESS))

    })
   .catch(error => {

     console.log("Oops: ", error)
     const txData = {
        code: "500",
        summary: `${UserConfig.resetFailure}`,
        time: `${dateText}`
     }
     dispatch(write({data: txData})(TxActionTypes.TX_FAILURE))
   })
  }
}
```

**Reply by rijkvanzanten** [reply]
I noticed that all requests in that network tab are 204/200, so I'm pretty sure that error isn't actually thrown by Directus. Seeing you're doing other stuff in those `.then` blocks in the promise, there's a non-zero chance that that's actually throwing the error, which is caught in the same "top level" catch 👍🏻

**Reply by glowkeeper** [reply]
That's why I showed that `console.log("did I get here?")`, though (which you can see gets fired) - that's pretty much the only thing that's doing in that first `.then`. Or am I misunderstanding you?

**Reply by rijkvanzanten** [reply]
This bit `dispatch(write({data: txData})(TxActionTypes.TX_SUCCESS))` can also throw an error that causes that `.catch`

**Reply by glowkeeper** [reply]
Ah, I see - Nah - that's just writing data to an internal `redux` store...no chance that's causing that `.catch`...
